jerry-ext/module/module.c

```diff
@@ -0,0 +1,225 @@
+jerryx_module_instance_insert (jerryx_module_instance_t *instance_p, /**< item to insert */
+                               jerryx_module_instance_t **root_pp) /**< pointer to root (as it may change) */
+{
+  if (*root_pp)
+  {
+    instance_p->next_p = (*root_pp);
+  }
+  (*root_pp) = instance_p;
+} /* jerryx_module_instance_insert */
+
+/*
+ * Find an item in a linked list.
+ *
+ * @return the requested item, or NULL if not found
+ */
+static jerryx_module_instance_t *
+jerryx_module_instance_find (jerryx_module_instance_t *root_p, /**< the root of the list */
+                             const jerry_char_t *name) /**< the key item to find */
+{
+  for (;root_p != NULL; root_p = root_p->next_p)
+  {
+    if (!strcmp ((char *) name, ((char *) root_p->name)))
+    {
+      break;
+    }
+  }
+
+  return root_p;
+} /* jerryx_module_instance_find */
+
+/**
+ * Deeply delete a linked list.
+ */
+static void
+jerryx_module_instances_free (jerryx_module_instance_t *root_p) /**< root of the list */
+{
+  jerryx_module_instance_t *next_p = NULL;
+  while (root_p)
+  {
+    next_p = root_p->next_p;
+    jerry_release_value (root_p->exports);
+    jmem_heap_free_block (root_p, sizeof (jerryx_module_instance_t));
+    root_p = next_p;
+  }
+} /* jerryx_module_instances_free */
+
+/**
+ * Deinitialize the module manager extension.
+ */
+static void
+jerryx_module_manager_deinit (void *user_data_p) /**< context pointer to deinitialize */
+{
+  jerryx_module_instances_free (*((jerryx_module_instance_t **) (user_data_p)));
+} /* jerryx_module_manager_deinit */
+
+/**
+ * Declare the context data manager for modules.
+ */
+static const jerry_context_data_manager_t jerryx_module_manager =
+{
+  .init_cb = NULL,
+  .deinit_cb = jerryx_module_manager_deinit,
+  .bytes_needed = sizeof (jerryx_module_instance_t *)
+};
+
+/**
+ * Declare the linker section where module definitions are stored.
+ */
+JERRYX_SECTION_DECLARE (jxm_modules, jerryx_module_t)
+
+/**
+ * Declare and define the default module resolver - one which examines what modules are defined in the above linker
+ * section and loads one that matches the requested name, caching the result for subsequent requests using the context
+ * data mechanism.
+ */
+JERRYX_MODULE_RESOLVER (native_resolver)
+{
+  int index;
+  jerry_value_t ret = 0;
+  const jerryx_module_t *module_p = NULL;
+  jerryx_module_instance_t *instance_p = NULL;
+  jerryx_module_instance_t **instances_pp =
+  (jerryx_module_instance_t **) jerry_get_context_data (&jerryx_module_manager);
+
+  instance_p = jerryx_module_instance_find (*instances_pp, name);
+
+  if (instance_p)
+  {
+    return instance_p->exports;
+  }
+
+  for JERRYX_SECTION_ITERATE (jxm_modules, index, module_p)
+  {
+    if (!strcmp ((char *) module_p->name, (char *) name))
+    {
+      if (module_p->on_resolve)
+      {
+        ret = module_p->on_resolve ();
+        instance_p = (jerryx_module_instance_t *) jmem_heap_alloc_block (sizeof (jerryx_module_instance_t));
```

Comment:
That's fine with me.

I also wonder whether it would be useful if all the resolvers would share the same cache map, instead of each resolver having to keep its own piece of the cache (just move the caching mechanism a level up, into `jerryx_module_resolve`). That way each resolver can just focus on resolving and not have to worry about caching.